README.md
# JavaScript Encrypted Data Vault Client _(edv-client)_

[![Build Status](https://travis-ci.org/digitalbazaar/edv-client.png?branch=master)](https://travis-ci.org/digitalbazaar/edv-client)

> A JavaScript library for Web and node.js apps for interfacing with a remote
> Encrypted Data Vault server

## Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [API](#api)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

This library provides a client that Web and node.js apps can use to interface
with remote Encrypted Data Vault (EDV) servers.

It consists of one main class:

1. `EdvClient` - instances provide a CRUD (+ find) interface to a specific
  configured Encrypted Data Vault server and ensure appropriate database
  indexes are set up. Static methods allow for the creation of EDVs with a
  remote storage service, e.g.
    [Encrypted Data Vault storage server](https://github.com/digitalbazaar/bedrock-data-hub-storage).

## Install

- Browsers and Node.js 18+ are supported.
- [Streams API][] required. Older browsers must use a polyfill.
- [Web Crypto API][] required. Older browsers must use a polyfill.

To install from NPM:

```
npm install @digitalbazaar/edv-client
```

To install locally (for development):

```
git clone https://github.com/digitalbazaar/edv-client.git
cd edv-client
npm install
```

## Usage

### Creating and registering an Encrypted Data Vault (EDV)

First, create a key agreement key and an HMAC (hash-based message authentication
code) key for encrypting your documents and blinding any indexed attributes in
them. This requires creating some cryptographic key material which can be done
locally or via a KMS system. The current example shows using a KMS system
(TODO: show a simpler local example):

```js
import {CapabilityAgent, KeystoreAgent, KmsClient} from 'webkms-client';
import {EdvClient} from 'edv-client';

```
Although Encrypted Data Vaults are not bound to any particular key management
system, we recommend that you set up a Key Management Service using an
implementation such as
[`webkms-switch`](https://github.com/digitalbazaar/webkms-switch)
which you can connect to using
[`webkms-client`](https://github.com/digitalbazaar/webkms-client).

Optional:

```js
// create a CapabilityAgent (for invoking zcaps)
const capabilityAgent = await CapabilityAgent.fromSecret({secret, handle});

// create a keystore and an agent for working with it
// the baseUrl can be set to a dev API or production API
const kmsBaseUrl = `${config.server.baseUri}/kms`;
const keystore = KmsClient.createKeystore({
  // the url for the keystore is configurable
  url: `${kmsBaseUrl}/keystores`,
  config: {
    // on init the sequence must be 0
    sequence: 0,
    controller: capabilityAgent.id,
    invoker: capabilityAgent.id,
    // this allows the capabilityAgent to delegate zCaps
    delegator: capabilityAgent.id
  },
  /**
   * optional `httpsAgent`,
   * usually not applicable for front-end (you may use axios),
   * for back-end use cases a nodejs `https.Agent`
   * may be used to allow the use of self signed certificates using
   * the `rejectUnauthorized: false` flag in the contructor.
  */
  httpsAgent
});
const keystoreAgent = new KeystoreAgent({keystore, capabilityAgent});

// use the keystore agent to create key agreement and HMAC keys
const keyAgreementKey = await keystoreAgent.generateKey({type: 'keyAgreement'});
const hmac = await keystoreAgent.generateKey({type: 'hmac'});
```

Now you can create and register a new EDV configuration:

```js
// TODO: explain EDV service must be able to authenticate user
const controller = 'account id goes here';

const config = {
  // on init the sequence must be 0 and is required
  sequence: 0,
  controller,
  // TODO: Explain what 'referenceId' is
  referenceId: 'primary',
  keyAgreementKey: {id: keyAgreementKey.id, type: keyAgreementKey.type},
  hmac: {id: hmac.id, type: hmac.type}
};

// sends a POST request to the remote service to create an EDV
const remoteConfig = await EdvClient.createEdv({config});

// connect to the new EDV via a `EdvClient`
const client = new EdvClient({id: remoteConfig.id, keyAgreementKey, hmac});
// to only use fips-compliant key agreement and encryption algorithms:
const client = new EdvClient(
  {id: remoteConfig.id, keyAgreementKey, hmac, cipherVersion: 'fips'});
```

You can specify a url when you create and register a new EDV configuration:

```js
// TODO: explain EDV service must be able to authenticate user
const controller = 'account id (or DID if using zcaps) goes here';

const config = {
  // on init the sequence must be 0 and is required
  sequence: 0,
  controller,
  // TODO: Explain what 'referenceId' is
  referenceId: 'primary',
  keyAgreementKey: {id: keyAgreementKey.id, type: keyAgreementKey.type},
  hmac: {id: hmac.id, type: hmac.type}
};

// sends a POST request to the remote service to create an EDV
const remoteConfig = await EdvClient.createEdv({
  url: 'https://server.example/edvs',
  config,
  // must pass `invocationSigner` and optional `capability` if `controller`
  // is a DID
  /* invocationSigner, capability */
});

// connect to the new EDV via a `EdvClient`
const client = new EdvClient({id: remoteConfig.id, keyAgreementKey, hmac});
```


### Loading a saved EDV config

If you have previously registered an EDV config (via `createEdv()`),
and you know its `id`, you can fetch its config via `get()`:

```js
// registered config
const {id} = await EdvClient.createEdv({config});

// later, it can be fetched via the id
const remoteConfig = await EdvClient.getConfig({id});

// connect to the existing EDV via an `EdvClient` instance
const client = new EdvClient({id: remoteConfig.id, keyAgreementKey, hmac});
```

If you know a controller/`accountId` but do not know a specific EDV `id`, you
can create a client for an EDV by a controller-scoped custom `referenceId`:

```js
// get the account's 'primary' EDV config to connect to the EDV
// note that a referenceId can be any string but must be unique per controller
const config = await EdvClient.findConfig(
  {controller: accountId, referenceId: 'primary'});
const client = new EdvClient({id: config.id, keyAgreementKey, hmac});
```

### Using a EdvClient instance for document storage

See the API section below.

## API

### `EdvClient`

#### `constructor`

#### `insert`

#### `get`

#### `update`

#### `delete`

#### `find`

#### `ensureIndex`

#### `updateIndex`

## Contribute

Please follow the existing code style.

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[BSD-3-Clause](LICENSE.md) © Digital Bazaar

[Streams API]: https://developer.mozilla.org/en-US/docs/Web/API/Streams_API
[Web Crypto API]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API


CHANGELOG.md
# @digitalbazaar/edv-client ChangeLog

## 16.1.0 - 2023-11-07

### Added
- Add support for using fips-compliant algorithms for key agreement via
  `@digitalbazaar/minimal-cipher@6`.
- Add new `cipherVersion` parameter to control whether the cipher version
  is "recommended" or "fips". The default (and previous only option)
  remains "recommended".

## 16.0.1 - 2023-11-05

### Changed
- Fix linting errors and update test dependencies. No API changes.

## 16.0.0 - 2023-09-14

### Changed
- **BREAKING**: Drop support for Node.js < 18.
- Use `@digitalbazaar/http-client@4` which requires Node.js 18+.
- Use `canonicalize@2`. This version adds error handling for illegal numeric
  values: Infinite and NaN.

## 15.0.2 - 2022-08-18

### Fixed
- Move `base64url-universal` from `devDependencies` to `dependencies`.

## 15.0.1 - 2022-08-14

### Fixed
- Use `@digitalbazaar/minimal-cipher@5.1.1` to get chacha bug fix.

## 15.0.0 - 2022-06-09

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Require Node.js >=14.
- **BREAKING**: Use `globalThis` for browser crypto and streams.
- **BREAKING**: Require Web Crypto API. Older browsers and Node.js 14 users
  need to install an appropriate polyfill.
- **BREAKING**: Require Streams API. Older browsers and Node.js <18 users need
  to install an appropriate polyfill.
- Update dependencies.
- Lint module.

## 14.1.1 - 2022-05-06

### Fixed
- Simplify prewarm cache algorithm.

## 14.1.0 - 2022-05-06

### Added
- Added cache for blinding attributes and code to prewarm it.

### Changed
- Use `Map` and promises instead of `object` for internal cached key
  resolver when processing streams.

## 14.0.0 - 2022-05-05

### Changed
- **BREAKING**: Blinded attributes (aka encrypted attributes/indexes) are
  now calculated differently and are incompatible with previous versions. A
  private `_migrate()` tool has been provided on `EdvClient` that can do
  basic migration from the old version to the new version, but is very
  limited to small queries (< 1000 docs) and must be run in a context that
  ensures unique constraints are not violated but concurrent writes that
  are external to the migration process. A more robust tool will only be
  written if it is found to be necessary.

## 13.0.0 - 2022-03-01

### Changed
- **BREAKING**: Better future proof zcap endpoints by posting zcap
  revocations to `/zcaps/revocations` instead of just `/revocations`.

## 12.0.2 - 2022-02-24

### Fixed
- Ensure EDV ID is parsed from `capability` param.

## 12.0.1 - 2022-02-23

### Fixed
- Add missing `capability` param.

## 12.0.0 - 2022-02-23

### Changed
- **BREAKING**: Locally digest all values used to produce encrypted
  indexes by using sha-256 prior to HMAC'ing them. This increases
  privacy when using a WebKMS system. This approach to producing
  encrypted indexes will be incompatible with previously generated
  attributes.

## 11.3.2 - 2022-02-20

### Fixed
- Fix missing param defaults for `count`.

## 11.3.1 - 2022-02-16

### Fixed
- Fix use of zcap with EDV ID as its target in `find`.

## 11.3.0 - 2022-02-06

### Added
- Expose `generateId` on EdvClientCore instances as well as statically.

## 11.2.0 - 2022-02-05

### Added
- Support `limit` option in `query`. If `limit` is passed, it must be
  an integer >= 1 and <= 1000. An EDV server that understands it will
  limit the results to the given value and report back a boolean,
  `hasMore`, to indicate whether there are more results that match
  the query. There is presently no cursor value that can be passed
  in a future query to continue / skip previous results, this is
  deferred to a future version.

## 11.1.2 - 2022-02-02

### Fixed
- Fix bugs with compound indexes. Previously, when a document contained
  multiple values for an attribute (e.g., an array value was used) that
  participated in a compound index, it was possible for the earlier
  values to be erroneously dropped from the index. This bug has now
  been fixed.

## 11.1.1 - 2022-01-31

### Fixed
- Add missing files to package.

## 11.1.0 - 2022-01-31

### Added
- Add new underlying transport layer abstraction for EDV clients. This
  exposes a new class `EdvClientCore` that relies on a `Transport`
  interface to handle communicating with an EDV server / storage. This
  class can be extended by derived classes that provide transport
  implementations. `EdvClient` has been refactored to use an HTTPS
  transport layer -- but in a backwards compatible fashion. A future
  version may introduce breaking changes to further separate concerns
  more cleanly.
- Enable `capability` and `invocationSigner` to be passed to the
  `EdvClient` constructor. If these are passed, then they do not
  need to be passed when calling individual methods as the defaults
  will be used. Take care to ensure that the capability passed is
  usable in all methods that are to be called without passing in
  a specific capability (this is not significantly different concern
  from before, however, there is a subtle difference because of the
  reliance on a default value.

## 11.0.0 - 2022-01-11

### Changed
- **BREAKING**: Use http-signature-zcap-invoke@5. These changes
  include major breaking simplifications to ZCAP (zcap@7).
- **BREAKING**: Renamed package to `@digitalbazaar/edv-client`.

## 10.0.0 - 2021-09-01

### Changed
- **BREAKING**: All root zcaps use `urn:root:zcap:` prefix. Root zcaps
  for documents are the EDV root zcap where the controller resides. This new
  client version must be paired with a new EDV server, it is not compatible
  with an old version.
- **BREAKING**: `getConfig` is now an instance member function instead
  of a static class member function. It requires that a capability be
  invoked to fetch the EDV config.
- **BREAKING**: Use simplified zcap revocation model via `revokeCapability`.
  Now any party that has delegated a zcap may revoke it by calling
  `revokeCapability` with the revoked zcap without passing an additional

assert.js
/*!
 * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as base58 from 'base58-universal';

export function assert(variable, name, types) {
  if(!Array.isArray(types)) {
    types = [types];
  }
  const type = variable instanceof Uint8Array ? 'Uint8Array' : typeof variable;
  if(!types.includes(type) ||
    // an object must not falsey nor an array
    (type === 'object' && (!variable || Array.isArray(variable)))) {
    let article;
    if(types.length === 1) {
      article = types[0] === 'object' ? 'an' : 'a';
    }
    throw new TypeError(
      `"${name}" must be ${types.length === 1 ? article : 'one of'} ` +
      `${types.join(', ')}.`);
  }
}

export function assertDocument(doc) {
  assert(doc, 'doc', 'object');
  const {id, content, meta = {}, stream} = doc;
  if(id !== undefined) {
    assertDocId(doc.id);
  }
  assert(content, 'content', 'object');
  assert(meta, 'meta', 'object');
  if(stream !== undefined) {
    assert(stream, 'stream', 'object');
  }
}

export function assertDocId(id) {
  try {
    // verify ID is multibase base58-encoded 16 bytes
    const buf = base58.decode(id.substr(1));
    // multibase base58 (starts with 'z')
    // 128-bit random number, multibase encoded
    // 0x00 = identity tag, 0x10 = length (16 bytes) + 16 random bytes
    if(!(id.startsWith('z') &&
      buf.length === 18 && buf[0] === 0x00 && buf[1] === 0x10)) {
      throw new Error('Invalid document ID.');
    }
  } catch(e) {
    throw new Error(`Document ID "${id}" must be a multibase, base58-encoded ` +
      'array of 16 random bytes.');
  }
}

export function assertInvocationSigner(invocationSigner) {
  assert(invocationSigner, 'invocationSigner', 'object');
  const {id, sign} = invocationSigner;
  assert(id, 'invocationSigner.id', 'string');
  assert(sign, 'invocationSigner.sign', 'function');
}

export function assertTransport(transport) {
  assert(transport, 'transport', 'object');
}
